File: ts/checkpoint.ts
/**
 * Checkpoint reuse verdict — pure policy: no I/O, no timers, no clock.
 *
 * Extracted from a production planning pipeline's checkpoint-invalidation
 * module (issue fields and reason codes unchanged, except where noted).
 * Canonicalizes a value and digests it, and decides whether an artifact
 * recorded by an earlier run may be reused: its status must map to
 * `complete`, and its recorded contract revision, stage-config digest, and
 * dependency digests must equal what the caller expects now.
 *
 * Departures from the source, each so that every language computes the same
 * answer:
 * - `canonicalize` returns a halt for anything outside the spec's value model
 *   instead of serializing it however the host language happens to (the
 *   source turned `undefined` into invalid JSON and every class instance into
 *   `{}`). Like every verdict here it is a value, never a throw: the caller
 *   records it before it stops.
 * - A string is always a value: the source digested a bare string's raw
 *   bytes, so the text `{"a":1}` and the value `{a: 1}` collided. Callers
 *   that want a digest of raw text compute it themselves.
 * - `evaluateCheckpointArtifact` takes the artifact instead of reading a
 *   path, takes the caller's status vocabulary through `status_map`, compares
 *   dependencies in UTF-16 key order rather than insertion order, and returns
 *   issues without an `issue_id` — ids and file reads stay with the caller.
 *   The status reason is `artifact_status_not_reusable` (the source's
 *   `artifact_status_not_pass` named a vocabulary this spec does not have).
 * - A recorded `dependency_digests` that is not a plain object, and a key it
 *   only inherits (`constructor`), read as absent; the source indexed into
 *   lists, strings, and the prototype chain.
 * - The arguments are typed and an argument of another type is refused; the
 *   source read them by JavaScript falsiness, under which an expectation of
 *   `""` or `0` was no expectation, and ignored validation issues that were
 *   not a list. What the artifact *records* is still read leniently: it is
 *   data, written by whoever wrote it.
 * - A null field in a caller's validation issue is an absent one. The source
 *   laid it over the default, so a null `reason` came back as null.
 */
import { createHash } from "node:crypto";
import { checkFields } from "./contract.ts";

// ---------------------------------------------------------------- canonicalize

/** Largest integer every implementation represents exactly: 2^53 − 1. */
const MAX_SAFE_INTEGER = 9007199254740991;
const MAX_SAFE_BIGINT = 9007199254740991n;

/** Deepest nesting of lists and maps a digest input may have. A bound every
 * language can reach without exhausting its stack, so all of them halt at
 * the same depth instead of each crashing at its own. */
const MAX_DEPTH = 100;

export type DigestInputReason =
  | "digest_input_float"
  | "digest_input_int_range"
  | "digest_input_unsupported";

/** A value outside the digest value model: the halt's reason code, and a
 * message naming where the value sits (`$.a[2]`) for the person reading it.
 * The message is for people; only `halt` is part of the spec. */
export interface DigestInputHalt {
  halt: DigestInputReason;
  message: string;
}

/** Internal: unwinds the recursive encoder to the public boundary, where it
 * becomes a {@link DigestInputHalt} value. */
class DigestInputError extends Error {
  readonly reason: DigestInputReason;

  constructor(reason: DigestInputReason, message: string) {
    super(message);
    this.name = "DigestInputError";
    this.reason = reason;
  }
}

/** The seven characters with a two-character escape. Every other code point
 * below U+0020 is written `\u00xx` (lowercase hex); everything else, `/`,
 * U+007F, U+2028 and U+2029 included, is written as itself. This is exactly
 * what `JSON.stringify` does for well-formed strings — the table is spelled
 * out so a port without JSON.stringify can match it. */
const SHORT_ESCAPES: Readonly<Record<number, string>> = {
  0x08: "\\b",
  0x09: "\\t",
  0x0a: "\\n",
  0x0c: "\\f",
  0x0d: "\\r",
  0x22: '\\"',
  0x5c: "\\\\",
};

/**
 * Canonical form of a value: an RFC 8785 subset. Map keys are sorted by
 * UTF-16 code unit, there is no insignificant whitespace, integers are
 * written in decimal, and strings use the escape table above. Strings are
 * not Unicode-normalized: NFC and NFD spellings of the same text canonicalize
 * differently.
 *
 * Accepted: null, booleans, integers with |n| <= 2^53 − 1 whether stored as
 * a number or a bigint (so `1.0` is the integer 1 — the spec follows the
 * value, not the storage), strings of Unicode scalar values, and arrays and
 * plain objects nested at most 100 deep. Anything else returns a
 * {@link DigestInputHalt}.
 */
export function canonicalize(value: unknown): { canonical: string } | DigestInputHalt {
  try {
    return { canonical: encodeValue(value, "$", 0) };
  } catch (error) {
    if (error instanceof DigestInputError) return { halt: error.reason, message: error.message };
    throw error;
  }
}

/** `sha256:` + lowercase hex SHA-256 of the canonical form's UTF-8 bytes —
 * the same value `printf '%s' "$canonical" | sha256sum` prints — or the
 * halt {@link canonicalize} returned. */
export function checkpointDigest(value: unknown): { digest: string } | DigestInputHalt {
  const result = canonicalize(value);
  if ("halt" in result) return result;
  return { digest: `sha256:${createHash("sha256").update(result.canonical, "utf8").digest("hex")}` };
}

function encodeValue(value: unknown, at: string, depth: number): string {
  if (value === null) return "null";
  if (value === true) return "true";
  if (value === false) return "false";
  if (typeof value === "number") return encodeNumber(value, at);
  if (typeof value === "bigint") return encodeBigInt(value, at);
  if (typeof value === "string") return encodeString(value, at);
  if ((Array.isArray(value) || isPlainObject(value)) && depth >= MAX_DEPTH) {
    throw new DigestInputError(
      "digest_input_unsupported",
      `${at}: nested deeper than ${MAX_DEPTH} lists and maps`,
    );
  }
  if (Array.isArray(value)) {
    const parts: string[] = [];
    // An index loop, not map(): map() skips holes, and a hole must be
    // rejected like any other undefined rather than silently dropped.
    for (let index = 0; index < value.length; index += 1) {
      parts.push(encodeValue(value[index], `${at}[${index}]`, depth + 1));
    }
    return `[${parts.join(",")}]`;
  }
  if (isPlainObject(value)) {
    if (Object.getOwnPropertySymbols(value).length > 0) {
      throw new DigestInputError("digest_input_unsupported", `${at}: map has a symbol key`);
    }
    // The default sort compares UTF-16 code units, which is the spec's order.
    const keys = Object.keys(value).sort();
    const parts = keys.map(
      (key) => `${encodeString(key, `${at} key`)}:${encodeValue(value[key], `${at}.${key}`, depth + 1)}`,
    );
    return `{${parts.join(",")}}`;
  }
  throw new DigestInputError(
    "digest_input_unsupported",
    `${at}: ${describe(value)} is outside the digest value model`,
  );
}

function encodeNumber(value: number, at: string): string {
  if (!Number.isFinite(value) || !Number.isInteger(value)) {
    throw new DigestInputError(
      "digest_input_float",
      `${at}: ${String(value)} is not an integer; render it as a string if it belongs in a digest`,
    );
  }
  if (Math.abs(value) > MAX_SAFE_INTEGER) {
    throw new DigestInputError(
      "digest_input_int_range",
      `${at}: ${String(value)} is outside ±(2^53 − 1)`,
    );
  }
  // String(-0) is "0": zero is written without a sign.
  return String(value);
}

function encodeBigInt(value: bigint, at: string): string {
  if (value > MAX_SAFE_BIGINT || value < -MAX_SAFE_BIGINT) {
    throw new DigestInputError(
      "digest_input_int_range",
      `${at}: ${value.toString()} is outside ±(2^53 − 1)`,
    );
  }
  return value.toString();
}

function encodeString(value: string, at: string): string {
  let out = '"';
  for (let index = 0; index < value.length; index += 1) {
    const unit = value.charCodeAt(index);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const next = index + 1 < value.length ? value.charCodeAt(index + 1) : 0;
      if (next >= 0xdc00 && next <= 0xdfff) {
        out += String.fromCharCode(unit, next);
        index += 1;
        continue;
      }
    }
    if (unit >= 0xd800 && unit <= 0xdfff) {
      throw new DigestInputError(
        "digest_input_unsupported",
        `${at}: string has an unpaired surrogate at index ${index}`,
      );
    }
    const escaped = SHORT_ESCAPES[unit];
    if (escaped !== undefined) out += escaped;
    else if (unit < 0x20) out += `\\u${unit.toString(16).padStart(4, "0")}`;
    else out += String.fromCharCode(unit);
  }
  return `${out}"`;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  if (typeof value !== "object" || value === null) return false;
  const proto = Object.getPrototypeOf(value);
  return proto === Object.prototype || proto === null;
}

function describe(value: unknown): string {
  if (value === undefined) return "undefined";
  if (typeof value === "object" && value !== null) {
    return `an instance of ${value.constructor?.name ?? "an unknown class"}`;
  }
  return `a ${typeof value}`;
}

// ------------------------------------------------------------ reuse verdict

/** The status vocabulary every artifact is read through. Only `complete` is
 * reusable. */
export type ArtifactStatus = "complete" | "partial" | "failed" | "blocked";

/** Used when the caller passes no statusMap: the vocabulary maps to itself. */
const IDENTITY_STATUS_MAP: Readonly<Record<string, ArtifactStatus>> = {
  complete: "complete",
  partial: "partial",
  failed: "failed",
  blocked: "blocked",
};

export interface CheckpointIssue {
  stage_id: string;
  status: "valid" | "missing" | "invalid" | "unknown_contract";
  reason: string;
  required_resume_from_stage: string | null;
  subject_ref: string | null;
  [detail: string]: unknown;
}

/** The arguments, under the names the fixtures and every port use. Null and
 * not given are the same thing: absent. */
export interface EvaluateCheckpointArtifactArgs {
  stage_id: string;
  /** How issues refer to the artifact (a repo-relative path, a key). */
  subject_ref?: string | null;
  /** The artifact as recorded; absent when it does not exist. */
  artifact?: Readonly<Record<string, unknown>> | null;
  /** What the caller expects now. Absent: not checked. Any string is
   * compared, the empty one included. */
  expected_contract_revision?: string | null;
  expected_stage_config_digest?: string | null;
  expected_dependency_digests?: Readonly<Record<string, string>> | null;
  /** Where a rerun must start when this artifact is not reusable; defaults
   * to stage_id. */
  required_resume_from_stage?: string | null;
  /** Issues the caller's own validation found; each becomes an `invalid`
   * issue, with its own fields laid over the defaults. */
  validation_issues?: readonly Readonly<Record<string, unknown>>[] | null;
  /** The caller's status values mapped onto the vocabulary. Replaces the
   * default identity map; a status it does not name is not reusable. */
  status_map?: Readonly<Record<string, ArtifactStatus>> | null;
}

const ARGUMENT_FIELDS = [
  "stage_id",
  "subject_ref",
  "artifact",
  "expected_contract_revision",
  "expected_stage_config_digest",
  "expected_dependency_digests",
  "required_resume_from_stage",
  "validation_issues",
  "status_map",
];

/** An optional argument: null when absent, itself when it is what the
 * contract says, refused otherwise. */
function optional<T>(value: unknown, holds: (value: unknown) => value is T, what: string): T | null {
  if (value === null || value === undefined) return null;
  if (holds(value)) return value;
  throw new TypeError(`${what} is outside the contract`);
}

const isText = (value: unknown): value is string => typeof value === "string";
const isArtifactStatus = (value: unknown): value is ArtifactStatus =>
  isText(value) && Object.hasOwn(IDENTITY_STATUS_MAP, value);
const mapOf =
  <T>(holds: (value: unknown) => value is T) =>
  (value: unknown): value is Record<string, T> =>
    isPlainObject(value) && Object.values(value).every(holds);
const isListOfMaps = (value: unknown): value is Record<string, unknown>[] =>
  Array.isArray(value) && value.every(isPlainObject);

/**
 * Decide whether a recorded artifact may be reused. Returns every issue
 * found, in this order: status, contract revision, stage-config digest,
 * dependency digests (by UTF-16 key order), the caller's validation issues.
 * With none, returns one `valid` issue, so the result is never empty.
 */
export function evaluateCheckpointArtifact(args: EvaluateCheckpointArtifactArgs): CheckpointIssue[] {
  checkFields(args, ARGUMENT_FIELDS, "checkpoint arguments");
  const stageId: unknown = args.stage_id;
  if (!isText(stageId)) throw new TypeError("stage_id must be a string");
  const subjectRef = optional(args.subject_ref, isText, "subject_ref");
  const requiredResumeFromStage =
    optional(args.required_resume_from_stage, isText, "required_resume_from_stage") ?? stageId;
  const expectedRevision = optional(args.expected_contract_revision, isText, "expected_contract_revision");
  const expectedConfig = optional(args.expected_stage_config_digest, isText, "expected_stage_config_digest");
  const expectedDependencies =
    optional(args.expected_dependency_digests, mapOf(isText), "expected_dependency_digests") ?? {};
  const statusMap = optional(args.status_map, mapOf(isArtifactStatus), "status_map") ?? IDENTITY_STATUS_MAP;
  const validationIssues = optional(args.validation_issues, isListOfMaps, "validation_issues") ?? [];
  const artifact = optional(args.artifact, isPlainObject, "artifact");
  const base = (
    status: CheckpointIssue["status"],
    reason: string,
    details: Record<string, unknown> = {},
  ): CheckpointIssue => ({
    stage_id: stageId,
    status,
    reason,
    required_resume_from_stage: requiredResumeFromStage,
    subject_ref: subjectRef,
    ...details,
  });

  if (artifact === null) return [base("missing", "artifact_missing")];

  const issues: CheckpointIssue[] = [];
  const status = artifact.status;
  if (!status) {
    issues.push(base("invalid", "artifact_status_missing"));
  } else if (resolveStatus(status, statusMap) !== "complete") {
    issues.push(base("invalid", "artifact_status_not_reusable", { actual_status: status }));
  }

  if (expectedRevision !== null && !artifact.contract_revision) {
    issues.push(
      base("unknown_contract", "contract_revision_missing", {
        expected_contract_revision: expectedRevision,
      }),
    );
  } else if (expectedRevision !== null && artifact.contract_revision !== expectedRevision) {
    issues.push(
      base("invalid", "contract_revision_mismatch", {
        expected_contract_revision: expectedRevision,
        actual_contract_revision: artifact.contract_revision,
      }),
    );
  }

  if (expectedConfig !== null && artifact.stage_config_digest !== expectedConfig) {
    issues.push(
      base("invalid", "stage_config_digest_mismatch", {
        expected_stage_config_digest: expectedConfig,
        actual_stage_config_digest: artifact.stage_config_digest ?? null,
      }),
    );
  }

  const recordedDependencies = artifact.dependency_digests;
  for (const dependencyId of Object.keys(expectedDependencies).sort()) {
    const expectedDigest = expectedDependencies[dependencyId];
    const actualDigest =
      isPlainObject(recordedDependencies) && Object.hasOwn(recordedDependencies, dependencyId)
        ? (recordedDependencies[dependencyId] ?? null)
        : null;
    if (actualDigest !== expectedDigest) {
      issues.push(
        base("invalid", "dependency_digest_mismatch", {
          dependency_id: dependencyId,
          expected_digest: expectedDigest,
          actual_digest: actualDigest,
        }),
      );
    }
  }

  for (const validationIssue of validationIssues) {
    // A null field is an absent one: the default stands where there is one.
    const given = Object.entries(validationIssue).filter(([, field]) => field !== null && field !== undefined);
    issues.push({ ...base("invalid", "validation_issue"), ...Object.fromEntries(given) });
  }

  if (issues.length === 0) {
    return [{ ...base("valid", "checkpoint_valid"), required_resume_from_stage: null }];
  }
  return issues;
}

function resolveStatus(
  status: unknown,
  statusMap: Readonly<Record<string, ArtifactStatus>>,
): ArtifactStatus | null {
  // Own keys only: a status of "constructor" must not find Object's.
  if (typeof status !== "string" || !Object.hasOwn(statusMap, status)) return null;
  return statusMap[status] ?? null;
}

File: go/checkpoint.go
package haltrule

// Checkpoint - canonical form and digest of a value, and a verdict on whether
// an artifact an earlier run recorded may be reused.
//
// The arguments are typed, and an argument of another type cannot be built: a
// nil pointer or a nil map is absent, and nothing else is. What the artifact
// records is data, written by whoever wrote it, and is read leniently instead.

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
	"unicode/utf16"
	"unicode/utf8"
)

// maxSafeInteger is the largest integer every implementation represents
// exactly: 2^53 - 1.
const maxSafeInteger = 1<<53 - 1

// maxDepth is the deepest nesting of lists and maps a digest input may have. A
// bound every language reaches without exhausting its stack, so all of them
// halt at the same depth instead of each crashing at its own.
const maxDepth = 100

// DigestInputHalt is a value outside the digest value model: the halt's reason
// code, and a message naming where the value sits. The message is for people;
// only Halt is part of the spec.
type DigestInputHalt struct {
	Halt    string
	Message string
}

const (
	digestInputFloat       = "digest_input_float"
	digestInputIntRange    = "digest_input_int_range"
	digestInputUnsupported = "digest_input_unsupported"
)

func halted(reason, message string) *DigestInputHalt {
	return &DigestInputHalt{Halt: reason, Message: message}
}

// Canonicalize is the canonical form of a value: an RFC 8785 subset. Map keys
// are sorted by UTF-16 code unit, there is no insignificant whitespace,
// integers are written in decimal, and strings carry the spec's escapes.
// Strings are not Unicode-normalized.
func Canonicalize(value Value) (string, *DigestInputHalt) {
	return encodeValue(value, "$", 0)
}

// CheckpointDigest is "sha256:" and the lowercase hex SHA-256 of the canonical
// form's UTF-8 bytes, or the halt Canonicalize returned.
func CheckpointDigest(value Value) (string, *DigestInputHalt) {
	canonical, halt := Canonicalize(value)
	if halt != nil {
		return "", halt
	}
	sum := sha256.Sum256([]byte(canonical))
	return "sha256:" + hex.EncodeToString(sum[:]), nil
}

func encodeValue(value Value, at string, depth int) (string, *DigestInputHalt) {
	switch held := value.(type) {
	case nil:
		return "null", nil
	case Bool:
		if held {
			return "true", nil
		}
		return "false", nil
	case Int:
		if held > maxSafeInteger || held < -maxSafeInteger {
			return "", halted(digestInputIntRange, fmt.Sprintf("%s: %d is outside +/-(2^53 - 1)", at, int64(held)))
		}
		return strconv.FormatInt(int64(held), 10), nil
	case Float:
		number := float64(held)
		if math.IsNaN(number) || math.IsInf(number, 0) || number != math.Trunc(number) {
			return "", halted(digestInputFloat, fmt.Sprintf(
				"%s: %v is not an integer; render it as a string if it belongs in a digest", at, number))
		}
		if number > maxSafeInteger || number < -maxSafeInteger {
			return "", halted(digestInputIntRange, fmt.Sprintf("%s: %v is outside +/-(2^53 - 1)", at, number))
		}
		// A negative zero is written "0", as every zero is.
		return strconv.FormatInt(int64(number), 10), nil
	case String:
		return encodeString(string(held), at)
	case List:
		if depth >= maxDepth {
			return "", halted(digestInputUnsupported, fmt.Sprintf("%s: nested deeper than %d lists and maps", at, maxDepth))
		}
		parts := make([]string, 0, len(held))
		for index, item := range held {
			part, halt := encodeValue(item, fmt.Sprintf("%s[%d]", at, index), depth+1)
			if halt != nil {
				return "", halt
			}
			parts = append(parts, part)
		}
		return "[" + strings.Join(parts, ",") + "]", nil
	case Map:
		if depth >= maxDepth {
			return "", halted(digestInputUnsupported, fmt.Sprintf("%s: nested deeper than %d lists and maps", at, maxDepth))
		}
		parts := make([]string, 0, len(held))
		for _, key := range sortedKeys(held) {
			written, halt := encodeString(key, at+" key")
			if halt != nil {
				return "", halt
			}
			part, halt := encodeValue(held[key], fmt.Sprintf("%s.%s", at, key), depth+1)
			if halt != nil {
				return "", halt
			}
			parts = append(parts, written+":"+part)
		}
		return "{" + strings.Join(parts, ",") + "}", nil
	}
	// Value is closed, so there is no other kind to meet.
	return "", halted(digestInputUnsupported, at+": outside the digest value model")
}

// sortedKeys orders map keys by UTF-16 code unit, which is the spec's order
// and not Go's own, because they disagree above U+FFFF.
func sortedKeys(held Map) []string {
	keys := make([]string, 0, len(held))
	for key := range held {
		keys = append(keys, key)
	}
	// Go hands out a map's keys in a different order every run, so they are put
	// in one order before being put in the spec's. The answer is the same
	// either way; what this settles is the work done to reach it.
	sort.Strings(keys)
	sort.SliceStable(keys, func(left, right int) bool { return lessUTF16(keys[left], keys[right]) })
	return keys
}

func lessUTF16(left, right string) bool {
	leftUnits, rightUnits := utf16.Encode([]rune(left)), utf16.Encode([]rune(right))
	for index := 0; index < len(leftUnits) && index < len(rightUnits); index++ {
		if leftUnits[index] != rightUnits[index] {
			return leftUnits[index] < rightUnits[index]
		}
	}
	return len(leftUnits) < len(rightUnits)
}

// encodeString writes a string of Unicode scalar values, quoted, with exactly
// seven two-character escapes, every other code point below U+0020 as \u00xx
// in lowercase hex, and everything else as itself. A Go string that is not
// valid UTF-8 is not made of scalar values - an unpaired surrogate has no
// UTF-8 spelling - so it halts.
func encodeString(value, at string) (string, *DigestInputHalt) {
	if !utf8.ValidString(value) {
		return "", halted(digestInputUnsupported, at+": string is not made of Unicode scalar values")
	}
	var out strings.Builder
	out.WriteByte('"')
	for _, character := range value {
		switch character {
		case '"':
			out.WriteString(`\"`)
		case '\\':
			out.WriteString(`\\`)
		case '\b':
			out.WriteString(`\b`)
		case '\f':
			out.WriteString(`\f`)
		case '\n':
			out.WriteString(`\n`)
		case '\r':
			out.WriteString(`\r`)
		case '\t':
			out.WriteString(`\t`)
		default:
			if character < 0x20 {
				fmt.Fprintf(&out, `\u%04x`, character)
			} else {
				out.WriteRune(character)
			}
		}
	}
	out.WriteByte('"')
	return out.String(), nil
}

// ------------------------------------------------------------ reuse verdict

// ArtifactStatus is the vocabulary every artifact is read through. Only
// Complete is reusable.
type ArtifactStatus string

const (
	Complete ArtifactStatus = "complete"
	Partial  ArtifactStatus = "partial"
	Failed   ArtifactStatus = "failed"
	Blocked  ArtifactStatus = "blocked"
)

var identityStatusMap = map[string]ArtifactStatus{
	"complete": Complete, "partial": Partial, "failed": Failed, "blocked": Blocked,
}

// Issue is one finding about an artifact: stage_id, status, reason,
// required_resume_from_stage and subject_ref, and what its reason adds. A
// caller's own issue may lay any of them over the defaults, so it is a map.
type Issue map[string]Value

// CheckpointArgs is what the caller expects of an artifact now. A nil pointer
// or a nil map is absent; an empty map is present.
type CheckpointArgs struct {
	StageID    string
	SubjectRef *string
	// Artifact is what was recorded; nil when nothing was.
	Artifact Map
	// An absent expectation is not checked; any string is compared, the empty
	// one included.
	ExpectedContractRevision  *string
	ExpectedStageConfigDigest *string
	ExpectedDependencyDigests map[string]string
	// RequiredResumeFromStage defaults to StageID.
	RequiredResumeFromStage *string
	ValidationIssues        []Map
	// StatusMap replaces the default map, which sends each word of the
	// vocabulary to itself; a status it does not name is not reusable.
	StatusMap map[string]ArtifactStatus
}

func text(value *string) Value {
	if value == nil {
		return nil
	}
	return String(*value)
}

// EvaluateCheckpointArtifact decides whether a recorded artifact may be
// reused. It answers every issue found, in this order: status, contract
// revision, stage-config digest, dependency digests by UTF-16 key order, the
// caller's validation issues. With none it answers one valid issue, so the
// answer is never empty. It refuses a status map that answers outside the
// vocabulary.
func EvaluateCheckpointArtifact(args CheckpointArgs) ([]Issue, error) {
	statusMap := identityStatusMap
	if args.StatusMap != nil {
		for named, mapped := range args.StatusMap {
			if _, known := identityStatusMap[string(mapped)]; !known {
				return nil, fmt.Errorf("status_map sends %q outside the vocabulary, to %q", named, mapped)
			}
		}
		statusMap = args.StatusMap
	}
	resume := args.StageID
	if args.RequiredResumeFromStage != nil {
		resume = *args.RequiredResumeFromStage
	}
	base := func(status, reason string) Issue {
		return Issue{
			"stage_id":                   String(args.StageID),
			"status":                     String(status),
			"reason":                     String(reason),
			"required_resume_from_stage": String(resume),
			"subject_ref":                text(args.SubjectRef),
		}
	}

	if args.Artifact == nil {
		return []Issue{base("missing", "artifact_missing")}, nil
	}

	var issues []Issue
	status := args.Artifact["status"]
	if jsFalsy(status) {
		issues = append(issues, base("invalid", "artifact_status_missing"))
	} else if resolveStatus(status, statusMap) != Complete {
		issue := base("invalid", "artifact_status_not_reusable")
		issue["actual_status"] = status
		issues = append(issues, issue)
	}

	revision := args.Artifact["contract_revision"]
	if args.ExpectedContractRevision != nil && jsFalsy(revision) {
		issue := base("unknown_contract", "contract_revision_missing")
		issue["expected_contract_revision"] = String(*args.ExpectedContractRevision)
		issues = append(issues, issue)
	} else if args.ExpectedContractRevision != nil && !sameText(revision, *args.ExpectedContractRevision) {
		issue := base("invalid", "contract_revision_mismatch")
		issue["expected_contract_revision"] = String(*args.ExpectedContractRevision)
		issue["actual_contract_revision"] = revision
		issues = append(issues, issue)
	}

	config := args.Artifact["stage_config_digest"]
	if args.ExpectedStageConfigDigest != nil && !sameText(config, *args.ExpectedStageConfigDigest) {
		issue := base("invalid", "stage_config_digest_mismatch")
		issue["expected_stage_config_digest"] = String(*args.ExpectedStageConfigDigest)
		issue["actual_stage_config_digest"] = config
		issues = append(issues, issue)
	}

	recorded, recordedIsMap := args.Artifact["dependency_digests"].(Map)
	expectedIDs := make([]string, 0, len(args.ExpectedDependencyDigests))
	for id := range args.ExpectedDependencyDigests {
		expectedIDs = append(expectedIDs, id)
	}
	sort.Slice(expectedIDs, func(left, right int) bool { return lessUTF16(expectedIDs[left], expectedIDs[right]) })
	for _, id := range expectedIDs {
		expected := args.ExpectedDependencyDigests[id]
		var actual Value
		if recordedIsMap {
			actual = recorded[id]
		}
		if !sameText(actual, expected) {
			issue := base("invalid", "dependency_digest_mismatch")
			issue["dependency_id"] = String(id)
			issue["expected_digest"] = String(expected)
			issue["actual_digest"] = actual
			issues = append(issues, issue)
		}
	}

	for _, given := range args.ValidationIssues {
		issue := base("invalid", "validation_issue")
		// A null field is an absent one: the default stands where there is one.
		for key, field := range given {
			if field != nil {
				issue[key] = field
			}
		}
		issues = append(issues, issue)
	}

	if len(issues) == 0 {
		valid := base("valid", "checkpoint_valid")
		valid["required_resume_from_stage"] = nil
		return []Issue{valid}, nil
	}
	return issues, nil
}

func resolveStatus(status Value, statusMap map[string]ArtifactStatus) ArtifactStatus {
	named, isText := status.(String)
	if !isText {
		return ""
	}
	return statusMap[string(named)]
}

// jsFalsy is what "absent" means of a recorded value: null, false, 0, NaN and
// the empty string. An empty list or map is present.
func jsFalsy(value Value) bool {
	switch held := value.(type) {
	case nil:
		return true
	case Bool:
		return !bool(held)
	case Int:
		return held == 0
	case Float:
		return held == 0 || math.IsNaN(float64(held))
	case String:
		return held == ""
	}
	return false
}

// sameText is what equality means between a recorded value and an expectation:
// the same string, so a recorded 1 is not "1" and a recorded ["v2"] is not "v2".
func sameText(recorded Value, expected string) bool {
	held, isText := recorded.(String)
	return isText && string(held) == expected
}
